### PR TITLE
CoinbasePRO: Fix websocket subscription errors 

### PR DESCRIPTION
**Pull request by gloriousCode**
# PR Description
Coinbase does not allow parent productid subscriptions anymore. This PR puts the productIDs in the individual subscriptions

Before 😭 :
```
CoinbasePro websocket connection: sending message to websocket {Type:subscribe ProductIDs:[ETH-USD GMT-USDT MPL-USD NEAR-USDT CHZ-GBP MASK-EUR ATA-USD LINK-EUR CBETH-ETH USDT-USD] Channels:[{Name:heartbeat ProductIDs:[]} {Name:level2_batch ProductIDs:[]} {Name:ticker ProductIDs:[]} {Name:matches ProductIDs:[]}] Signature: Key: Passphrase: Timestamp:}

...

[DEBUG]	 | WEBSOCKET | 12/12/2023 14:07:07 | CoinbasePro websocket connection: message received: {"type":"error","message":"Failed to subscribe","reason":" is not a valid product"}
[DEBUG]	 | WEBSOCKET | 12/12/2023 14:07:07 | CoinbasePro websocket connection: message received: {"type":"error","message":"Failed to subscribe","reason":" is not a valid product"}
[DEBUG]	 | WEBSOCKET | 12/12/2023 14:07:07 | CoinbasePro websocket connection: message received: {"type":"error","message":"Failed to subscribe","reason":" is not a valid product"}
[DEBUG]	 | WEBSOCKET | 12/12/2023 14:07:07 | CoinbasePro websocket connection: message received: {"type":"error","message":"Failed to subscribe","reason":" is not a valid product"}
[DEBUG]	 | WEBSOCKET | 12/12/2023 14:07:07 | CoinbasePro websocket connection: message received: {"type":"subscriptions","channels":[]}
[ERROR]	 | WEBSOCKET | 12/12/2023 14:07:07 | exchange CoinbasePro websocket error - {"type":"error","message":"Failed to subscribe","reason":" is not a valid product"}
[ERROR]	 | WEBSOCKET | 12/12/2023 14:07:07 | exchange CoinbasePro websocket error - {"type":"error","message":"Failed to subscribe","reason":" is not a valid product"}
[ERROR]	 | WEBSOCKET | 12/12/2023 14:07:07 | exchange CoinbasePro websocket error - {"type":"error","message":"Failed to subscribe","reason":" is not a valid product"}
[ERROR]	 | WEBSOCKET | 12/12/2023 14:07:07 | exchange CoinbasePro websocket error - {"type":"error","message":"Failed to subscribe","reason":" is not a valid product"}

```

After 🌈 :
```
CoinbasePro websocket connection: sending message to websocket {Type:subscribe ProductIDs:[] Channels:[{Name:heartbeat ProductIDs:[ETH-USD GMT-USDT MPL-USD NEAR-USDT CHZ-GBP MASK-EUR ATA-USD LINK-EUR CBETH-ETH USDT-USD ETH-USD GMT-USDT MPL-USD NEAR-USDT CHZ-GBP MASK-EUR ATA-USD LINK-EUR CBETH-ETH USDT-USD ETH-USD GMT-USDT MPL-USD NEAR-USDT CHZ-GBP MASK-EUR ATA-USD LINK-EUR CBETH-ETH USDT-USD ETH-USD GMT-USDT MPL-USD NEAR-USDT CHZ-GBP MASK-EUR ATA-USD LINK-EUR CBETH-ETH USDT-USD]} {Name:level2_batch ProductIDs:[ETH-USD GMT-USDT MPL-USD NEAR-USDT CHZ-GBP MASK-EUR ATA-USD LINK-EUR CBETH-ETH USDT-USD ETH-USD GMT-USDT MPL-USD NEAR-USDT CHZ-GBP MASK-EUR ATA-USD LINK-EUR CBETH-ETH USDT-USD ETH-USD GMT-USDT MPL-USD NEAR-USDT CHZ-GBP MASK-EUR ATA-USD LINK-EUR CBETH-ETH USDT-USD ETH-USD GMT-USDT MPL-USD NEAR-USDT CHZ-GBP MASK-EUR ATA-USD LINK-EUR CBETH-ETH USDT-USD]} {Name:ticker ProductIDs:[ETH-USD GMT-USDT MPL-USD NEAR-USDT CHZ-GBP MASK-EUR ATA-USD LINK-EUR CBETH-ETH USDT-USD ETH-USD GMT-USDT MPL-USD NEAR-USDT CHZ-GBP MASK-EUR ATA-USD LINK-EUR CBETH-ETH USDT-USD ETH-USD GMT-USDT MPL-USD NEAR-USDT CHZ-GBP MASK-EUR ATA-USD LINK-EUR CBETH-ETH USDT-USD ETH-USD GMT-USDT MPL-USD NEAR-USDT CHZ-GBP MASK-EUR ATA-USD LINK-EUR CBETH-ETH USDT-USD]} {Name:matches ProductIDs:[ETH-USD GMT-USDT MPL-USD NEAR-USDT CHZ-GBP MASK-EUR ATA-USD LINK-EUR CBETH-ETH USDT-USD ETH-USD GMT-USDT MPL-USD NEAR-USDT CHZ-GBP MASK-EUR ATA-USD LINK-EUR CBETH-ETH USDT-USD ETH-USD GMT-USDT MPL-USD NEAR-USDT CHZ-GBP MASK-EUR ATA-USD LINK-EUR CBETH-ETH USDT-USD ETH-USD GMT-USDT MPL-USD NEAR-USDT CHZ-GBP MASK-EUR ATA-USD LINK-EUR CBETH-ETH USDT-USD]}] Signature: Key: Passphrase: Timestamp:}

...

[DEBUG]	 | WEBSOCKET | 12/12/2023 13:56:55 | CoinbasePro websocket connection: message received: {"type":"subscriptions","channels":[{"name":"level2_50","product_ids":["NEAR-USDT","CBETH-ETH","MPL-USD","CHZ-GBP","USDT-USD","ETH-USD","LINK-EUR","MASK-EUR","ATA-USD","GMT-USDT"]},{"name":"ticker","product_ids":["ATA-USD","CHZ-GBP","CBETH-ETH","MASK-EUR","NEAR-USDT","LINK-EUR","USDT-USD","ETH-USD","MPL-USD","GMT-USDT"]},{"name":"matches","product_ids":["ETH-USD","NEAR-USDT","LINK-EUR","ATA-USD","CHZ-GBP","GMT-USDT","MASK-EUR","USDT-USD","CBETH-ETH","MPL-USD"]},{"name":"heartbeat","product_ids":["ATA-USD","CHZ-GBP","LINK-EUR","MPL-USD","USDT-USD","MASK-EUR","GMT-USDT","NEAR-USDT","CBETH-ETH","ETH-USD"]}]}
[DEBUG]	 | WEBSOCKET | 12/12/2023 13:56:55 | CoinbasePro websocket connection: message received: {"type":"snapshot","product_id":"CHZ-GBP","asks":[["0.0668","20856.3"],["0.0669","89883.9"],["0.0670","89159.6"],["0.0671","20.0"],["0.0672","20.0"],["0.0673","20.0"],["0.0674","20.0"],["0.0675","20.0"],["0.0676","20.0"],["0.0677","20.0"],["0.0678","20.0"],["0.0679","20.0"],["0.0680","205.0"],["0.0681","20.0"],["0.0682","20.0"],["0.0683","20.0"],["0.0684","20.0"],["0.0685","20.0"],["0.0686","20.0"],["0.0687","20.0"],["0.0688","20.0"],["0.0689","20.0"],["0.0690","20.0"],["0.0691","20.0"],["0.0692","20.0"],["0.0693","20.0"],["0.0694","20.0"],["0.0695","20.0"],["0.0696","20.0"],["0.0697","20.0"],["0.0698","20.0"],["0.0699","20.0"],["0.0700","20.0"],["0.0701","20.0"],["0.0702","20.0"],["0.0703","20.0"],["0.0704","20.0"],["0.0705","20.0"],["0.0706","20.0"],["0.0707","20.0"],["0.0708","20.0"],["0.0709","20.0"],["0.0710","20.0"],["0.0711","20.0"],["0.0712","20.0"],["0.0713","20.0"],["0.0714","20.0"],["0.0715","20.0"],["0.0716","20.0"],["0.0717","20.0"],["0.0718","20.0"],["0.0719","20.0"],["0.0720","20.0"],["0.0721","20.0"],["0.0722","20.0"],["0.0723","20.0"],["0.0724","20.0"],["0.0725","20.0"],["0.0726","20.0"],["0.0727","20.0"],["0.0728","20.0"],["0.0729","20.0"],["0.0730","20.0"],["0.0731","20.0"],["0.0732","20.0"],["0.0733","20.0"],["0.0734","20.0"],["0.0735","1142.6"],["0.0736","20.0"],["0.0737","20.0"],["0.0738","20.0"],["0.0739","20.0"],["0.0740","20.0"],["0.0741","20.0"],["0.0742","20.0"],["0.0743","20.0"],["0.0744","20.0"],["0.0745","20.0"],["0.0746","20.0"],["0.0747","20.0"],["0.0748","20.0"],["0.0749","20.0"],["0.0750","158.6"],["0.0751","20.0"],["0.0752","8760.7"],["0.0753","20.0"],["0.0754","20.0"],["0.0755","20.0"],["0.0756","20.0"],["0.0757","20.0"],["0.0758","20.0"],["0.0759","20.0"],["0.0760","20.0"],["0.0761","20.0"],["0.0762","20.0"],["0.0763","20.0"],["0.0764","20.0"],["0.0765","20.0"],["0.0766","20.0"],["0.0767","20.0"],["0.0768","20.0"],["0.0769","20.0"],["0.0770","20.0"],["0.0771","20.0"],["0.0772","20.0"],["0.0773","20.0"],["0.0774","20.0"],["0.0775","20.0"],["0.0776","20.0"],["0.0800","1000.0"],["0.0850","135.2"],["0.0870","6436.9"],["0.0890","834.6"],["0.0900","4.0"],["0.0950","331.8"],["0.0962","4924.2"],["0.0980","4936.5"],["0.1100","5802.0"],["0.1122","59705.8"],["0.1150","2638.9"],["0.1153","95.9"],["0.1156","120.7"],["0.1179","88.9"],["0.1200","18821.4"],["0.1250","2585.5"],["0.1255","260.1"],["0.1300","16270.3"],["0.1389","507.8"],["0.1400","10957.5"],["0.1429","148.6"],["0.1450","118.4"],["0.1455","105.0"],["0.1467","118.4"],["0.1476","167.1"],["0.1500","245.6"],["0.1530","167.1"],["0.1600","210.3"],["0.1680","1597.8"],["0.1700","70.5"],["0.1747","2100.0"],["0.1766","83.6"],["0.1800","223.2"],["0.1900","1000.0"],["0.1950","14561.5"],["0.1980","781.2"],["0.2000","34963.7"],["0.2088","147.2"],["0.2094","100.7"],["0.2100","5010.0"],["0.2143","126.5"],["0.2145","261.1"],["0.2150","1699.7"],["0.2182","1740.0"],["0.2200","6976.2"],["0.2226","331.3"],["0.2265","135.2"],["0.2270","126.9"],["0.2334","12251.6"],["0.2400","1166.0"],["0.2425","171.4"],["0.2499","35.4"],["0.2500","1550.7"],["0.2513","30.0"],["0.2527","41.5"],["0.2533","100.0"],["0.2589","121.2"],["0.2599","34.5"],["0.2600","3553.2"],["0.2650","135.8"],["0.2700","296.2"],["0.2756","115.1"],["0.2760","62.2"],["0.2800","68.0"],["0.2900","5282.6"],["0.2948","90.3"],["0.3000","1445.8"],["0.3020","10.0"],["0.3184","200.0"],["0.3330","696.0"],["0.3400","500.0"],["0.3500","38115.2"],["0.3750","300.0"],["0.3800","2146.7"],["0.3990","139.9"],["0.4000","75467.4"],["0.4030","10.0"],["0.4163","1200.0"],["0.4400","400.0"],["0.4500","311.9"],["0.4600","200.0"],["0.4800","195.0"],["0.4951","202.0"],["0.5000","531.0"],["0.5447","400.0"],["0.6000","45914.1"],["0.6300","120.3"],["0.6400","8.8"],["0.6500","10540.4"],["0.7000","8.9"],["0.7500","11.4"],["0.9500","13874.3"],["0.9700","1000.0"],["0.9800","2246.0"],["1.0000","35606.5"],["1.5000","142.2"],["1.8000","8.9"],["2.1350","41.9"],["2.5000","20.0"],["3.0000","176.3"],["4.2000","99.2"],["5.0000","44.8"],["7.2000","4.3"],["10.0000","92.0"],["10.2173","45.8"],["15.0000","3.8"],["19.8500","4118.6"],["20.0000","5.0"],["20.1386","100.0"],["25.0000","3.8"],["45.7700","5.0"]],"bids":[["0.0667","20837.9"],["0.0666","45039.8"],["0.0665","62708.7"],["0.0664","74753.9"],["0.0663","20.0"],["0.0662","20.0"],["0.0661","20.0"],["0.0660","20.0"],["0.0659","20.0"],["0.0658","20.0"],["0.0657","20.0"],["0.0656","20.0"],["0.0655","20.0"],["0.0654","20.0"],["0.0653","20.0"],["0.0652","20.0"],["0.0651","20.0"],["0.0650","20.0"],["0.0649","20.0"],["0.0648","20.0"],["0.0647","20.0"],["0.0646","20.0"],["0.0645","20.0"],["0.0644","20.0"],["0.0612","2500.0"],["0.0610","1314.7"],["0.0605","5054.3"],["0.0596","3498.5"],["0.0583","3558.6"],["0.0560","8875.0"],["0.0500","10198.8"],["0.0450","662.6"],["0.0440","2711.8"],["0.0434","20.0"],["0.0411","500.0"],["0.0400","5994.0"],["0.0260","7646.1"],["0.0250","199.3"],["0.0059","840.0"],["0.0050","380.0"],["0.0037","50.1"],["0.0001","6329940.0"]],"time":"2023-12-12T02:56:55.062139Z"}
```

## Type of change
- [x] Bug fix (non-breaking change which fixes an issue)

## How has this been tested

- [x] go test ./... -race
- [x] golangci-lint run
- [x] Running the application

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation and regenerated documentation via the documentation tool
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally and on Github Actions/AppVeyor with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: exchanges/coinbasepro/coinbasepro_websocket.go [hunk 1]
```diff
@@ -411,26 +411,29 @@ func (c *CoinbasePro) Subscribe(channelsToSubscribe []stream.ChannelSubscription
...
 		Type: "subscribe",
 	}
-
-subscriptions:
+	productIDs := make([]string, 0, len(channelsToSubscribe))
 	for i := range channelsToSubscribe {
 		p := channelsToSubscribe[i].Currency.String()
-		if !common.StringDataCompare(subscribe.ProductIDs, p) && p != "" {
-			subscribe.ProductIDs = append(subscribe.ProductIDs, p)
+		if p != "" && !common.StringDataCompare(subscribe.ProductIDs, p) {
```
Comment (shazbert) [inline exchanges/coinbasepro/coinbasepro_websocket.go:417]: should compare to `productIDs` as this will keep appending

**Review by shazbert — CHANGES_REQUESTED** [review]
Just one thing from me, don't have keys to test auth.

File: exchanges/coinbasepro/coinbasepro_websocket.go (hunk 1 above)
Comment (gloriousCode) [reply exchanges/coinbasepro/coinbasepro_websocket.go]: Good catch 🧤 I have done that + added the update to unsubscibes too. The way I tested it was to call unsubscribe once subscriptions were done 🌞

**Review by shazbert — APPROVED** [review]
Changes tACK. Nice work!

**Review by thrasher- — APPROVED** [review]
tACK, nice!

image

NOTE: Since this exchange is currently being updated to the latest "Coinbase Advanced Trading"  in a separate PR, will request websocket test coverage on that one instead of this one.

**Pull request merged**